### PR TITLE
Add support for copying ahb without cpu read mask

### LINKED ISSUE #1627: [replay] Background is not drawn during the loading screen of some apps
There are multiple android games where during the loading screen in the replay the background will just be some green color, which I assume is the clear color. Happens with trim and without it.

### PR DESCRIPTION
**Pull request by ziga-lunarg**
Closes #1627

### CODE REVIEW

**Review by mikes-lunarg — COMMENTED** [review]
Looks really good, just a couple of questions

File: framework/encode/vulkan_capture_manager.h [hunk 1]
```diff
@@ -1,6 +1,6 @@
 /*
  ** Copyright (c) 2018-2021 Valve Corporation
- ** Copyright (c) 2018-2023 LunarG, Inc.
+ ** Copyright (c) 2018-2025 LunarG, Inc.
```
Comment (antonio-lunarg) [inline framework/encode/vulkan_capture_manager.h:3]: No other change in this file?

File: framework/decode/vulkan_replay_consumer_base.cpp [hunk 2]
```diff
@@ -381,53 +381,526 @@ void VulkanReplayConsumerBase::ProcessFillMemoryCommand(uint64_t       memory_id
...
+                bufferCreateInfo.flags       = 0u;
+                bufferCreateInfo.size        = ahb_size;
+                bufferCreateInfo.usage       = VK_BUFFER_USAGE_TRANSFER_DST_BIT | VK_BUFFER_USAGE_TRANSFER_SRC_BIT;
+                bufferCreateInfo.sharingMode = VK_SHARING_MODE_EXCLUSIVE;
+                bufferCreateInfo.queueFamilyIndexCount = 0u;
+                bufferCreateInfo.pQueueFamilyIndices   = nullptr;
+
+                VkBuffer ahbDataBuffer;
+                if (vk_result == VK_SUCCESS)
+                    vk_result = device_table->CreateBuffer(device, &bufferCreateInfo, nullptr, &ahbDataBuffer);
```
[the remark this answers is not in the document]
Comment (ziga-lunarg) [reply framework/decode/vulkan_replay_consumer_base.cpp:483]: I put all vulkan api calls inside `if (vk_result == VK_SUCCESS)`, not sure what I would check if I put the check immediately after each call. The point is to not proceed with later api calls if a previous has failed. And I cannot return, because at the end I have to cleanup all objects that were successfully created

File: framework/decode/vulkan_replay_consumer_base.cpp [hunk 3]
```diff
@@ -381,53 +381,526 @@ void VulkanReplayConsumerBase::ProcessFillMemoryCommand(uint64_t       memory_id
...
+                device_table->DestroyCommandPool(device, commandPool, nullptr);
+                device_table->DestroyDescriptorPool(device, descriptorPool, nullptr);
+                device_table->DestroyPipeline(device, computePipeline, nullptr);
+                device_table->DestroyDescriptorSetLayout(device, descriptorSetLayout, nullptr);
+                device_table->DestroyPipelineLayout(device, pipelineLayout, nullptr);
+                device_table->DestroyShaderModule(device, computeShaderModule, nullptr);
+                device_table->FreeMemory(device, bufferMemory, nullptr);
+                device_table->FreeMemory(device, hostReadableBufferMemory, nullptr);
+                device_table->DestroyBuffer(device, ahbDataBuffer, nullptr);
+                device_table->DestroyBuffer(device, hostReadableBuffer, nullptr);
```
[the remark this answers is not in the document]
Comment (ziga-lunarg) [reply framework/decode/vulkan_replay_consumer_base.cpp:900]: Considering this does not seem to have a performance impact I thought it might be better to create these as needed

File: framework/decode/vulkan_replay_consumer_base.cpp [hunk 4]
```diff
@@ -381,53 +381,526 @@ void VulkanReplayConsumerBase::ProcessFillMemoryCommand(uint64_t       memory_id
...
+                bufferCreateInfo.pQueueFamilyIndices   = nullptr;
+
+                VkBuffer ahbDataBuffer;
+                if (vk_result == VK_SUCCESS)
+                    vk_result = device_table->CreateBuffer(device, &bufferCreateInfo, nullptr, &ahbDataBuffer);
+
+                VkImportAndroidHardwareBufferInfoANDROID importAHBInfo;
+                importAHBInfo.sType  = VK_STRUCTURE_TYPE_IMPORT_ANDROID_HARDWARE_BUFFER_INFO_ANDROID;
+                importAHBInfo.pNext  = nullptr;
+                importAHBInfo.buffer = buffer_info.hardware_buffer;
```
Comment (antonio-lunarg) [inline framework/decode/vulkan_replay_consumer_base.cpp:488]: I am afraid we can not do this with the current AHB flags: [VUID-VkMemoryAllocateInfo-pNext-02384](https://registry.khronos.org/vulkan/specs/latest/man/html/VkMemoryAllocateInfo.html#VUID-VkMemoryAllocateInfo-pNext-02384)

> If the parameters define an import operation and the external handle
type is
`VK_EXTERNAL_MEMORY_HANDLE_TYPE_ANDROID_HARDWARE_BUFFER_BIT_ANDROID`,
and the `pNext` chain does not include a
[VkMemoryDedicatedAllocateInfo](https://registry.khronos.org/vulkan/specs/latest/man/html/VkMemoryDedicatedAllocateInfo.html) structure or
[VkMemoryDedicatedAllocateInfo](https://registry.khronos.org/vulkan/specs/latest/man/html/VkMemoryDedicatedAllocateInfo.html)::image is
[VK_NULL_HANDLE](https://registry.khronos.org/vulkan/specs/latest/man/html/VK_NULL_HANDLE.html), the Android hardware buffer must have a
`AHardwareBuffer_Desc::format` of
`AHARDWAREBUFFER_FORMAT_BLOB` and a
`AHardwareBuffer_Desc::usage` that includes
`AHARDWAREBUFFER_USAGE_GPU_DATA_BUFFER`.

Thinking forward for an alternative solution, a viable alternative might be importing it using a VkImage. There are some restrictions there as well: [VUID-VkImageCreateInfo-pNext-02397](https://registry.khronos.org/vulkan/specs/latest/man/html/VkImageCreateInfo.html#VUID-VkImageCreateInfo-pNext-02397).

> If the pNext chain includes a [VkExternalFormatANDROID](https://registry.khronos.org/vulkan/specs/latest/man/html/VkExternalFormatANDROID.html) structure whose externalFormat member is not 0, usage must not include any usages except VK_IMAGE_USAGE_INPUT_ATTACHMENT_BIT, VK_IMAGE_USAGE_COLOR_ATTACHMENT_BIT, or VK_IMAGE_USAGE_SAMPLED_BIT

In other words, an AHB with externalFormat can not be copied directly to host visible memory via a transfer operation. It can only be resolved via attachment or sampled.
